client-core/src/service/key_service.rs

```diff
@@ -56,7 +92,152 @@ where
 
     /// Clears all storage
     pub fn clear(&self) -> Result<()> {
-        self.storage.clear(KEYSPACE)
+        self.storage.clear(KEYSPACE).expect("clear keyspace");
+        self.storage
+            .clear(KEYSPACE_HD)
+            .expect("clear keyspace for hd");
+        Ok(())
+    }
+
+    /// get random mnemonic
+    pub fn get_random_mnemonic(&self) -> String {
+        let mnemonic = Mnemonic::new(MnemonicType::Words24, Language::English);
+        mnemonic.clone().to_string()
+    }
+
+    /// is hd
+    pub fn is_hd_wallet(&self, name: &str, passphrase: &SecUtf8) -> bool {
+        let key = name.as_bytes();
+        let value = self.read_value(passphrase, &key[..]);
+        value.is_some()
+    }
+    /// generate seed from mnemonic
+    pub fn generate_seed(&self, mnemonic: &str, name: &str, passphrase: &SecUtf8) -> Result<()> {
+        debug!("hdwallet generate seed={}", mnemonic);
+        let mnemonic = Mnemonic::from_phrase(&mnemonic.to_string(), Language::English).unwrap();
+        let seed = Seed::new(&mnemonic, "");
+        self.storage
+            .set_secure(KEYSPACE_HD, name, seed.as_bytes().into(), passphrase)?;
+        debug!("hdwallet write seed success");
+        Ok(())
+    }
+
+    /// auto-matically generate staking, transfer addresses
+    /// with just one api call
+    #[allow(dead_code)]
+    pub fn auto_restore(
+        &self,
+        mnemonic: &str,
+        name: &str,
+        passphrase: &SecUtf8,
+        count: i32,
+    ) -> Result<()> {
+        self.generate_seed(mnemonic, name, passphrase)
+            .expect("auto restore");
+        let cointype = get_bip44_coin_type();
+        println!("coin type={}", cointype);
+        for index in 0..count {
+            let seed_bytes = self.storage.get_secure(KEYSPACE_HD, name, passphrase)?;
+            for account in 0..2 {
+                let extended = ExtendedPrivKey::derive(
+                    &seed_bytes.clone().unwrap()[..],
+                    format!("m/44'/{}'/{}'/0/{}", cointype, account, index).as_str(),
+                )
+                .unwrap();
+                let secret_key_bytes = extended.secret();
+
+                let private_key = PrivateKey::deserialize_from(&secret_key_bytes).unwrap();
+                let public_key = PublicKey::from(&private_key);
+
+                self.storage.set_secure(
+                    KEYSPACE,
+                    public_key.serialize(),
+                    private_key.serialize(),
+                    passphrase,
+                )?;
+            }
+        }
+        Ok(())
+    }
+
+    /// read value from db, if it's None, there value doesn't exist
+    pub fn read_value(&self, passphrase: &SecUtf8, key: &[u8]) -> Option<Vec<u8>> {
+        if let Ok(connected) = self.storage.get_secure(KEYSPACE_HD, key, passphrase) {
+            if let Some(value) = connected {
+                return Some(value.clone());
+            }
+        }
+        None
+    }
```

Comment:
Instead of returning `Option<Vec<u8>>`, return `Result<Option<Vec<u8>>>` and handle all the errors.